### DISCUSSION TITLE
Typescript: how to handle resolvers that return partial data?

### DISCUSSION THREAD

**Discussion by aryzing** [dotansimha/graphql-code-generator#4030] [General] [answered]
With Apollo, resolvers are allowed to return partial data, which is then fed into the object type resolver to fill in the rest. For example:

```js
const resolvers = {
  Query: {
    getHuman(_, { id }) {
      const human = readFromDb(id);
      // {
      //   id: id,
      //   age: 26,
      //   height: 123,
      // }
      return human;
    },
    Human: {
      friends({ id }) {
        const friends = friendsOf(id);
        // [{id: 'friend-1-id'}, {id: 'friend-2-id'}, ...]
        return friends;
      },
    },
  },
};
```

However, if the schema defines `getHuman: Human!`, then the `typescript-resolvers` will set the type of the `getHuman` resolver as a function that must return an object that has a `friends`,

```js
export type Human = {
   __typename?: 'Human';
  id: Scalars['ID'];
  age: Scalars['Number'];
  height: Scalars['Number'];
  friends: Array<Human>;
};
```

Of course, the `getHuman` resolver purposefully does not return the type above. We're relying on the `Human` definition to fill in the `friends` field.

At this point, I don't know how to proceed: TS is rightfully reporting a type mismatch. How do I make this work?

**Comment by ardatan** [comment, marked as answer]
Actually GraphQL in general allows you to return any kind of data in resolvers. So that's why we have `mappers` to let specify the types for those.
https://graphql-code-generator.com/docs/plugins/typescript-resolvers#mappers---overwrite-parents-and-resolved-values

**Reply by aryzing** [reply]
Spot on, thanks @ardatan !

**Reply by dotansimha** [reply]
@aryzing I've just wrote a short tutorial for using `mappers`, you might find this helpful :) https://the-guild.dev/blog/better-type-safety-for-resolvers-with-graphql-codegen

**Reply by loucadufault** [reply]
Any chance of making the tutorial usable again (the code examples are dead). Thanks!
